### ISSUE TITLE
`trait_newly_sealed` false positive

### ISSUE THREAD

**Issue by chrjabs** [obi1kenobi/cargo-semver-checks#1200]
### Which lint or lints are the issue

`trait_newly_sealed`

### Known issues that might be causing this

- [ ] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

I was unfortunately not able to produce a small example to reproduce this. I'm seeing the issue in the [RustSAT repository](https://github.com/chrjabs/rustsat), where running `cargo semver-checks` produces the output attached below, asking for a major bump that I am pretty sure isn't needed.

Things I have already checked:
- The trait in question _can_ be implemented from an external crate, so it is definitely not sealed
- I ran a `git bisect` in the repository, and the commit to cause the semver check to fail is [`0a245b7`](https://github.com/chrjabs/rustsat/commit/0a245b7ecc49eb3afb663808eb22599220680b74)
- I minimized the above failing commit to [`eb0o0dc`](https://github.com/chrjabs/rustsat/commit/eb090dcf121ffb4d195163be56ded7758a3d2606), which is a quite trivial commit that applied to the baseline causes the semver check to fail. The [`semvercheck-debugging`](https://github.com/chrjabs/rustsat/tree/semvercheck-debugging) branch can be used for reproducing it, it only contains the above commit after the last release.

### Actual Behaviour

The `trait_newly_sealed` lint fails and `cargo semver-checks` suggest a major bump, even though the trait is not sealed.

### Expected Behaviour

The lint should pass and a minor bump be suggested.

### Verbose Lint Output

```
->> cargo semver-checks --verbose
    Updating crates.io index
     Locking 63 packages to latest compatible versions
    Building rustsat v0.6.4 (current)
 Documenting rustsat v0.6.4 (/persist/home/christoph/Work/rustsat)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.99s
       Built [   4.056s] (current)
     Parsing rustsat v0.6.4 (current)
      Parsed [   0.084s] (current)
     Parsing rustsat v0.6.4 (baseline, cached)
      Parsed [   0.045s] (baseline)
    Checking rustsat v0.6.4 -> v0.6.4 (no change)
    Starting 148 checks, 0 unnecessary on 16 threads
        PASS [   0.001s]       major        attribute_proc_macro_missing
        PASS [   0.031s]       major        auto_trait_impl_removed
        PASS [   0.003s]       major        constructible_struct_adds_field
        PASS [   0.006s]       major        constructible_struct_adds_private_field
        PASS [   0.002s]       major        constructible_struct_changed_type
        PASS [   0.005s]       major        declarative_macro_missing
        PASS [   0.001s]       major        derive_helper_attr_removed
        PASS [   0.001s]       major        derive_proc_macro_missing
        PASS [   0.029s]       major        derive_trait_impl_removed
        PASS [   0.002s]       major        enum_discriminants_undefined_non_exhaustive_variant
        PASS [   0.002s]       major        enum_discriminants_undefined_non_unit_variant
        PASS [   0.002s]       major        enum_marked_non_exhaustive
        PASS [   0.002s]       major        enum_missing
        PASS [   0.002s]       minor        enum_must_use_added
        PASS [   0.003s]       major        enum_no_repr_variant_discriminant_changed
        PASS [   0.002s]       major        enum_now_doc_hidden
        PASS [   0.002s]       major        enum_repr_int_changed
        PASS [   0.002s]       major        enum_repr_int_removed
        PASS [   0.002s]       major        enum_repr_transparent_removed
        PASS [   0.003s]       major        enum_repr_variant_discriminant_changed
        PASS [   0.002s]       major        enum_struct_variant_field_added
        PASS [   0.006s]       major        enum_struct_variant_field_missing
        PASS [   0.003s]       major        enum_struct_variant_field_now_doc_hidden
        PASS [   0.003s]       major        enum_tuple_variant_changed_kind
        PASS [   0.004s]       major        enum_tuple_variant_field_added
        PASS [   0.003s]       major        enum_tuple_variant_field_missing
        PASS [   0.003s]       major        enum_tuple_variant_field_now_doc_hidden
        PASS [   0.003s]       major        enum_unit_variant_changed_kind
        PASS [   0.004s]       major        enum_variant_added
        PASS [   0.003s]       major        enum_variant_marked_non_exhaustive
        PASS [   0.003s]       major        enum_variant_missing
        PASS [   0.001s]       major        exported_function_changed_abi
        PASS [   0.001s]       major        feature_missing
        PASS [   0.000s]       major        feature_not_enabled_by_default
        PASS [   0.002s]       major        function_abi_no_longer_unwind
        PASS [   0.002s]       major        function_abi_now_unwind
        PASS [   0.002s]       major        function_changed_abi
        PASS [   0.002s]       major        function_const_removed
        PASS [   0.002s]       major        function_export_name_changed
        PASS [   0.002s]       major        function_like_proc_macro_missing
        PASS [   0.002s]       minor        function_marked_deprecated
        PASS [   0.002s]       major        function_missing
        PASS [   0.002s]       minor        function_must_use_added
        PASS [   0.002s]       major        function_now_doc_hidden
        PASS [   0.002s]       major        function_parameter_count_changed
        PASS [   0.002s]       major        function_requires_different_const_generic_params
        PASS [   0.003s]       major        function_requires_different_generic_type_params
        PASS [   0.002s]       major        function_unsafe_added
        PASS [   0.001s]       minor        global_value_marked_deprecated
        PASS [   0.002s]       major        inherent_associated_const_now_doc_hidden
        PASS [   0.005s]       major        inherent_associated_pub_const_missing
        PASS [   0.004s]       major        inherent_method_const_removed
        PASS [   0.020s]       major        inherent_method_missing
        PASS [   0.014s]       minor        inherent_method_must_use_added
        PASS [   0.019s]       major        inherent_method_now_doc_hidden
        PASS [   0.030s]       major        inherent_method_unsafe_added
        PASS [   0.001s]       minor        macro_marked_deprecated
        PASS [   0.001s]       major        macro_no_longer_exported
        PASS [   0.001s]       major        macro_now_doc_hidden
        PASS [   0.039s]       major        method_parameter_count_changed
        PASS [   0.026s]       major        method_requires_different_const_generic_params
        PASS [   0.032s]       major        method_requires_different_generic_type_params
        PASS [   0.002s]       major        module_missing
        PASS [   0.001s]       major        non_exhaustive_struct_changed_type
        PASS [   0.003s]       major        partial_ord_enum_struct_variant_fields_reordered
        PASS [   0.002s]       major        partial_ord_enum_variants_reordered
        PASS [   0.004s]       major        partial_ord_struct_fields_reordered
        PASS [   0.002s]       minor        proc_macro_marked_deprecated
        PASS [   0.001s]       major        proc_macro_now_doc_hidden
        PASS [   0.002s]       minor        pub_api_sealed_trait_became_unconditionally_sealed
        PASS [   0.002s]       minor        pub_api_sealed_trait_became_unsealed
        PASS [   0.002s]       major        pub_module_level_const_missing
        PASS [   0.002s]       major        pub_module_level_const_now_doc_hidden
        PASS [   0.002s]       major        pub_static_missing
        PASS [   0.001s]       major        pub_static_mut_now_immutable
        PASS [   0.001s]       major        pub_static_now_doc_hidden
        PASS [   0.001s]       major        pub_static_now_mutable
        PASS [   0.002s]       major        repr_c_enum_struct_variant_fields_reordered
        PASS [   0.002s]       major        repr_c_plain_struct_fields_reordered
        PASS [   0.002s]       major        repr_c_removed
        PASS [   0.002s]       major        repr_packed_added
        PASS [   0.001s]       major        repr_packed_removed
        PASS [   0.008s]       major        sized_impl_removed
        PASS [   0.001s]       major        static_became_unsafe
        PASS [   0.002s]       minor        struct_field_marked_deprecated
        PASS [   0.002s]       major        struct_marked_non_exhaustive
        PASS [   0.003s]       major        struct_missing
        PASS [   0.002s]       minor        struct_must_use_added
        PASS [   0.003s]       major        struct_now_doc_hidden
        PASS [   0.002s]       major        struct_pub_field_missing
        PASS [   0.002s]       major        struct_pub_field_now_doc_hidden
        PASS [   0.003s]       major        struct_repr_transparent_removed
        PASS [   0.004s]       major        struct_with_no_pub_fields_changed_type
        PASS [   0.004s]       major        struct_with_pub_fields_changed_type
        PASS [   0.006s]       major        trait_added_supertrait
        PASS [   0.002s]       major        trait_allows_fewer_const_generic_params
        PASS [   0.006s]       major        trait_allows_fewer_generic_type_params
        PASS [   0.002s]       major        trait_associated_const_added
        PASS [   0.002s]       major        trait_associated_const_default_removed
        PASS [   0.002s]       minor        trait_associated_const_marked_deprecated
        PASS [   0.001s]       major        trait_associated_const_now_doc_hidden
        PASS [   0.002s]       major        trait_associated_type_added
        PASS [   0.002s]       major        trait_associated_type_default_removed
        PASS [   0.002s]       minor        trait_associated_type_marked_deprecated
        PASS [   0.002s]       major        trait_associated_type_now_doc_hidden
        PASS [   0.001s]       minor        trait_marked_deprecated
        PASS [   0.002s]       major        trait_method_added
        PASS [   0.002s]       major        trait_method_default_impl_removed
        PASS [   0.002s]       minor        trait_method_marked_deprecated
        PASS [   0.003s]       major        trait_method_missing
        PASS [   0.003s]       major        trait_method_now_doc_hidden
        PASS [   0.006s]       major        trait_method_parameter_count_changed
        PASS [   0.004s]       major        trait_method_requires_different_const_generic_params
        PASS [   0.010s]       major        trait_method_requires_different_generic_type_params
        PASS [   0.003s]       major        trait_method_unsafe_added
        PASS [   0.002s]       major        trait_method_unsafe_removed
        PASS [   0.002s]       major        trait_mismatched_generic_lifetimes
        PASS [   0.002s]       major        trait_missing
        PASS [   0.002s]       minor        trait_must_use_added
        FAIL [   0.002s]       major        trait_newly_sealed
        PASS [   0.001s]       major        trait_no_longer_dyn_compatible
        PASS [   0.002s]       major        trait_now_doc_hidden
        PASS [   0.001s]       major        trait_removed_associated_constant
        PASS [   0.001s]       major        trait_removed_associated_type
        PASS [   0.001s]       major        trait_removed_supertrait
        PASS [   0.004s]       major        trait_requires_more_const_generic_params
        PASS [   0.005s]       major        trait_requires_more_generic_type_params
        PASS [   0.002s]       major        trait_unsafe_added
        PASS [   0.001s]       major        trait_unsafe_removed
        PASS [   0.002s]       major        tuple_struct_to_plain_struct
        PASS [   0.005s]       major        type_allows_fewer_const_generic_params
        PASS [   0.005s]       major        type_allows_fewer_generic_type_params
        PASS [   0.002s]       minor        type_associated_const_marked_deprecated
        PASS [   0.002s]       minor        type_marked_deprecated
        PASS [   0.005s]       minor        type_method_marked_deprecated
        PASS [   0.007s]       major        type_mismatched_generic_lifetimes
        PASS [   0.007s]       major        type_requires_more_const_generic_params
        PASS [   0.004s]       major        type_requires_more_generic_type_params
        PASS [   0.002s]       minor        unconditionally_sealed_trait_became_pub_api_sealed
        PASS [   0.001s]       minor        unconditionally_sealed_trait_became_unsealed
        PASS [   0.002s]       major        union_field_added_with_all_pub_fields
        PASS [   0.002s]       major        union_field_added_with_non_pub_fields
        PASS [   0.001s]       major        union_field_missing
        PASS [   0.001s]       major        union_missing
        PASS [   0.001s]       minor        union_must_use_added
        PASS [   0.001s]       major        union_now_doc_hidden
        PASS [   0.001s]       major        union_pub_field_now_doc_hidden
        PASS [   0.001s]       major        unit_struct_changed_kind
     Checked [   0.063s] 148 checks: 147 pass, 1 fail, 0 warn, 0 skip

--- failure trait_newly_sealed: pub trait became sealed ---

Description:
A publicly-visible trait became sealed, so downstream crates are no longer able to implement it
        ref: https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.40.0/src/lints/trait_newly_sealed.ron

Failed in:
  trait rustsat::encodings::card::EncodeIncremental in file /persist/home/christoph/Work/rustsat/src/encodings/card.rs:261

     Summary semver requires new major version: 1 major and 0 minor checks failed
    Finished [   4.890s] rustsat
```

#### Software version

cargo-semver-checks 0.40.0

#### Operating system

- OS: Linux (NixOS 24.11)
- Kernel: 6.6.58

#### Command-line

```bash
/nix/store/2fgll7dkpi5qk02ws8rl934h8z1hjljp-cargo-semver-checks-0.40.0/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.85.0 (d73d2caf9 2024-12-31)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

**Comment by chrjabs** [comment]
I just noticed something interesting. I have a GitHub action workflow on pull requests that also runs `cargo semver-checks`, but in that workflow the opposite is reported: `unconditionally_sealed_trait_became_unsealed`

https://github.com/chrjabs/rustsat/actions/runs/13951569828/job/39052185725#step:3:1

**Comment by obi1kenobi** [comment]
This is fascinating! Apologies for the trouble, this does indeed look like a bug.

I think I'll likely need to set aside a decent chunk of time to dig into this properly, so it might not happen before the weekend.

In the meantime, if the lints are in the way, you can temporarily disable them with `Cargo.toml` configuration by setting them to `"allow"`.

**Comment by chrjabs** [comment]
Let me know if I can help with anything else for reproducing this.

**Comment by obi1kenobi** [comment]
I have successfully reproduced the issue, and I'm extremely confused about it. Current thoughts: "the code is haunted!" 😆 

Thanks for bisecting and minimizing the repro, that'll help a lot.

**Comment by starovoid** [comment]
Hi, it looks like the same problem is occurring in `petgraph`: https://github.com/petgraph/petgraph/pull/764#issuecomment-2782207938
I can still implement the specified traits (`Create` and `Build`) in the downstream crate.

**Comment by obi1kenobi** [comment]
Apologies for the delay, I wasn't able to spend sufficient time on getting to the bottom of this issue. The Google Summer of Code candidate application deadline is in the next couple of days, and I've been a bit overwhelmed with everyone's questions :)

Hoping to have more time later this week!

**Comment by beeb** [comment]
Adding another example here, hope it helps!

https://github.com/beeb/lintspec/blob/e50f927681ba6d838e391b84d079a83ba2e4bf25/src/definitions/mod.rs#L34-L45

This public trait which was previously using the re-exported `slang_solidity::cst::TextRange` and now using the crate's own `TextRange` is considered sealed by cargo-semver-checks, but I'm able to implement the trait externally.

**Comment by obi1kenobi** [comment]
After debugging for most of the day today, I believe I have figured out the issues. There's more than one bug, unfortunately.

The issue pointed out by @chrjabs and @starovoid appears to be a bug in the cycle-tracking code deep inside the sealed trait analysis. I believe it misidentifies diamond DAGs as cycles due to a logic error. I have verified a fix against both of your repro cases, and I just want to cover the edge case with good test cases before merging the fix.

The issue reported by @beeb is much more challenging, and is due to our very limited ability to analyze `pub type` items. We have several more bugs related to `pub type`, but this is by far the most severe and it will be a difficult one to fix. I wrote out a sketch of some ideas [in a draft PR](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/825), but there's a long way to go before that PR can be merged.

To get some intuition about why `pub type` is full of edge cases, consider a case like `pub type Foo = Bar<Quux>` and try to figure out the rules about when `Foo, Bar, Quux` are public API or not (and whether they would cause trait-sealing), depending on whether any one or more of them are `#[doc(hidden)]`, private, or public but inside a private module and therefore non-importable. You'll quickly see that the relationship is very complex! For example, `Bar` could be non-importable by itself (so use of `Bar` in a trait method would cause the trait to be sealed), _except_ if the `Bar` in the trait in question is actually `Bar<Quux>` specifically — in which case that unseals the trait by allowing use of the type alias `Foo` instead.

Solving this problem properly will take hundreds of hours of work. We will eventually get there, but I don't have a timeline to offer. I'd encourage everyone interested in speeding up our tempo there to [become a GitHub Sponsor of the project](https://github.com/sponsors/obi1kenobi)!

**Comment by obi1kenobi** [comment]
I made #1254 to track the issue @beeb discovered, since its resolution will be different compared to the rest of the cases here.

I have published the fix in new `trustfall-rustdoc-adapter` versions in the v36-v45 range, and will close this issue shortly as soon as my `cargo update` PR updating to those versions is merged.

Thanks to everyone for contributing high-quality repros, and for your patience while I debugged this unfortunate and tricky problem.
